Accept tuple.0.0 as tuple indexing

dtolnay opened the pull request:
Previously rustc would reject `tuple.0.0`, requiring the programmer to write either `tuple.0 .0` or `(tuple.0).0` to work around the limitation because the `0.0` is emitted as one float token by the lexer.

This PR adjusts the lexer to produce <kbd>Ident</kbd><kbd>Dot</kbd><kbd>Integer</kbd><kbd>Dot</kbd><kbd>Integer</kbd> for the above, instead of <kbd>Ident</kbd><kbd>Dot</kbd><kbd>Float</kbd>, thereby making `tuple.0.0` work without a parser change.

---

- Relevant issue: #28257
- Tuple indexing RFC: https://github.com/rust-lang/rfcs/pull/184
- https://github.com/rust-lang/rfcs/pull/184#issuecomment-53498305 (2014) says Niko preferred not to have this at the time, but that we can consider supporting it down the line. https://github.com/rust-lang/rust/issues/28257#issuecomment-138858977 (2015) says Niko considers it may have been the wrong call to reject this originally.

---

r? @Centril

FYI @nikomatsakis since your ancient comments came up. :heart: 

FYI @petrochenkov and @matklad &mdash; I would be interested whether you are on board with a lexer change like this.

dtolnay committed:
Lex tuple.0.0 as tuple indexing

dtolnay committed:
Remove tuple-float-index ui test that now compiles

dtolnay committed:
Add test of tuple nested indexing

dtolnay committed:
Remove obsolete recovery of field access by float

dtolnay committed:
Test even deeper nested indexing

dtolnay committed:
Test a range, which is not nested indexing

Centril commented on src/librustc_parse/parser/expr.rs:
```diff
@@ -783,47 +780,6 @@ impl<'a> Parser<'a> {
         self.struct_span_err(self.token.span, &format!("unexpected token: `{}`", actual)).emit();
     }
 
-    fn recover_field_access_by_float_lit(
```
Happy to see this method go; it was ungreat to refactor in the past :) 🎉

Centril commented:
From the POV of someone who is quite familiar with `rustc_parse` but not with `rustc_lexer` (so I'm not an expert in the subject matter of this PR, and so the technical reviewer should be r? @petrochenkov) this change seems quite minimal and reasonable. While it somewhat complicates the spec it doesn't seem like it does so notably, and on the other hand it simplifies the compiler internals in practical terms (*diff is negative*).

Assuming @petrochenkov & @matklad are happy with this I would be happy to propose FCP merging this to the language team.

Though let's also nominate it for discussion today on our language team meeting.

matklad commented:
Am I correct that `tuple. 0.0` won't work? We should add this as a test. And probably `tuple./*i love special cases*/0.0` as well. 

Am I correct that this change will break this program?

```rust
macro_rules! m { (.$l:literal) => (); }
m!(.0.0);

fn main() {}
```

It should be added as a test as well. 

Another case I can think of is  `(((),),).000.000`. I think for this one behavior before and after would be the same, but better test than sorry. 

Generally, I see two different ways how we can achieve this behavior:

* tweak the lexing (like this PR does), using a rule like "it is not an FP literal, if preceded by dot"
* tweak the parsing (like the original recovery did), using a rule like "a field name can be an ident, and int, or a float with a specific text".

I personally would lean towards the second approach, as it will produce less "false positives". It means that the parser should be exposed to the *text* of the token, and not only to its lexical class, but it is already somewhat happens for contextual keywords (although this case is different, as the set of context keywords is finite, while the set of \d+.\d+ float literals is not).

Makes me wonder why we didn't choose `((),)._0` as a syntax for tuple fields in the first place :)

dtolnay committed:
Add tests in which the token really is a float

dtolnay committed:
Add test for tuple indexed with float in macro input

dtolnay committed:
Add test for errors triggered on parts of decomposed index

dtolnay commented:
> Am I correct that `tuple. 0.0` won't work? We should add this as a test. And probably `tuple./*i love special cases*/0.0` as well.

Added tests. These are now the way to write <kbd>Ident</kbd><kbd>Dot</kbd><kbd>Float</kbd> if that's what the programmer means, but that should be quite rare. It's better than requiring a workaround for <kbd>Ident</kbd><kbd>Dot</kbd><kbd>Integer</kbd><kbd>Dot</kbd><kbd>Integer</kbd>.

<br>

> Am I correct that this change will break this program?
> `macro_rules! m { (.$l:literal) => (); } m!(.0.0);`
> It should be added as a test as well.

Added test.

<br>

> Another case I can think of is `(((),),).000.000`. I think for this one behavior before and after would be the same, but better test than sorry.

Added test.

bjorn3 commented:
I think whitespace sensitivity is bad. This may also break the pretty printing of expressions as tokenstreams (`proc_macro::TokenStream::to_string()`) asd I believe the pretty printing inserts a space between every token.

dtolnay commented:
> Generally, I see two different ways how we can achieve this behavior:
>
> - tweak the lexing (like this PR does), using a rule like "it is not an FP literal, if preceded by dot"
> - tweak the parsing (like the original recovery did), using a rule like "a field name can be an ident, and int, or a float with a specific text".
>
> I personally would lean towards the second approach, as it will produce less "false positives".

I obviously lean toward the first approach. Some reasons:

- I do not want `$expr.$float` to work in a macro. In other words I really do see failure to recognize `tuple.0.0` as a lexer bug, not as "we should make floats a valid two-level index." There should not be floats involved in the first place!

- This implementation cleanly and correctly relexes the remainder of the content after the dot. For example in `tuple.0.0b0e0` we naturally get the lexer's existing handling of the "binary float literal is not supported" error on the `0b0e0` float literal. Matching this behavior in the parser after having treated `b0e0` as a "suffix" on a `0.0` float token would be more complicated and require duplicating the lexer's checks.

- I'm not sure what you mean by false positives. Could you give an example of a problematic false positive? A false positive would be something that is considered a tuple index when it shouldn't be. That seems unlikely to arise in practice.

In any case, if the lang team concludes that `tuple.0.0` is supposed to be accepted, I'm sure we can agree on one or the other for the implementation.

dtolnay commented:
> I think whitespace sensitivity is bad.

That's fine, but this PR is no more whitespace sensitive than `0.0` vs `0 . 0` is whitespace sensitive.

Furthermore, you'll love that this PR makes things *less whitespace sensitive*, such as in the case of `tuple.0.0` vs `tuple.0 .0` currently being whitespace sensitive but not after this PR. I think this PR is an improvement on the whitespace sensitivity axis.

<br>

> Also this may break the pretty printing of expressions as tokenstreams (`proc_macro::TokenStream::to_string()`) as I believe the pretty printing inserts a space between every token.

They would come out as `tuple . 0 . 0` is fine and expected.

petrochenkov commented:
This reminds me of our treatment of punctuation in proc macros.
`&&` is not `token::AndAnd`, but two tokens `token::And` where the first one has the "joint" flag.
Parser can then treat joint `&`s as a logical AND.

I remember some people (@eddyb at least?) regretted that we didn't break up floats in the same way as multi-character operators while proc macros were unstable, and didn't move the treatment of floats entirely to the parser.
(I believe this came up recently in some @rust-lang/wg-grammar discussion.)

In `rustc_parse` we haven't migrated to the proc macro model with simple joint and non-joint tokens yet (but we plan to).
So right now, when we want `token::And`, but encounter a `token::AndAnd`, we use [`fn break_and_eat`](https://github.com/rust-lang/rust/blob/8ff785011be6625e32afceee3a08e5cff7470feb/src/librustc_parse/parser/mod.rs#L552)  to break it into two tokens on the fly.
I think it would be reasonable to do the same thing with floats. Want a `token::Integer`, but encountered a `token::Float`? Break it into multiple tokens.
The @matklad's "second way" more or less.

(Perhaps later we'll be able to migrate floats to the "simple token" model internally and expose them as a single token only to proc macros for compatibility. Or maybe not, depending on convenience mostly.)

petrochenkov commented:
With the scheme from my comment above `$expr.$float` won't work because `$float` is not a float token, it's an "interpolated" token, it basically have invisible parentheses around it.

Centril commented:
We briefly discussed this on the language team meeting, but @petrochenkov's comment wasn't there yet, so I think we'll need some more discussion on the next one. :)

Centril commented:
We discussed this again in this week's language team meeting taking @petrochenkov's notes into account this time. Some conclusions from the meeting:

- Of those present, no one objected to the idea of accepting this in some fashion (though it will need to go through FCP to confirm and to take community input into consideration, as usual).

- Niko noted that if we were do to it by changing the lexer, as currently done in this PR (and as alluded to by @petrochenkov re. "compatibility"), there could be procedural macros observing the difference in the tokens they receive. It would be unlikely, but it would be breaking.

- We would like to see the suggested strategy (https://github.com/rust-lang/rust/pull/70420#issuecomment-604638130) of breaking float tokens when an integer is wanted explored so that we can compare them and make more informed decisions.

Dylan-DPC commented:
@dtolnay any updates?

dtolnay commented:
I haven't gotten a chance to try the alternative implementation suggested in https://github.com/rust-lang/rust/pull/70420#issuecomment-604638130. Maybe @Centril or someone else already familiar with the parser would like to give it a shot?

petrochenkov commented:
I can try implementing it this weekend.

petrochenkov commented:
The alternative implemenation - https://github.com/rust-lang/rust/pull/71322.

dtolnay commented:
Based on https://github.com/rust-lang/rust/pull/71322#issuecomment-622161421, closing in favor of #71322. Thanks @petrochenkov for the nicer approach and implementation!
